### PR TITLE
rustc: Make the `trait_map` of TyCtxt private

### PR DESCRIPTION
**Pull request by alexcrichton**
This map is calculated in resolve, but we want to be sure to track it for
incremental compliation. Hide it behind a query to get more refactorings later.

cc #44137

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -1997,3 +1999,13 @@ impl<T, R, E> InternIteratorElement<T, R> for Result<T, E> {
         Ok(f(&iter.collect::<Result<AccumulateVec<[_; 8]>, _>>()?))
     }
 }
+
+fn in_scope_traits<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, id: DefId)
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: I'm worried about the proliferation of providers, but I think making `trait_map` public to `rustc::ty` would be worse.

File: src/librustc/ty/context.rs [hunk 2]
```diff
@@ -1997,3 +1999,13 @@ impl<T, R, E> InternIteratorElement<T, R> for Result<T, E> {
...
+    -> Option<Rc<Vec<TraitCandidate>>>
+{
```
Comment (arielb1) [inline src/librustc/ty/context.rs]: I don't like the proliferation of `provide` methods, but I think making `trait_map` public in all of `rustc::ty` would be worse.

**Comment by arielb1** [comment]
Basically LGTM, but there's a travis blip and I suspect the primary use-case of `TraitMap` in https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/method/probe.rs#L664-L676 hasn't been updated.

**Comment by alexcrichton** [comment]
Ok I think I handled that case (still waiting on a local compile)

I also switched this over to being keyed on a `HirId` at @nikomatsakis's suggestion

**Comment by alexcrichton** [comment]
Ok I've also pushed up a commit which hides the `export_map` in a similar fashion.

re-r? @arielb1 

also cc @nikomatsakis

**Pull request merged**